### PR TITLE
Fix `no-descending-specificity` false positives for nested rules without declarations

### PR DESCRIPTION
**Pull request by romainmenke**
> Which issue, if any, is this issue related to?

Closes https://github.com/stylelint/stylelint/issues/7846

> Is there anything in the PR that needs further explanation?

No, it's self-explanatory.

### CODE REVIEW

File: lib/rules/no-descending-specificity/README.md
Comment (Mouvedia) [inline lib/rules/no-descending-specificity/README.md:1]: minor
fix

Check line 28 and 30.

File: lib/rules/no-descending-specificity/index.mjs [hunk 1]
```diff
@@ -151,6 +157,21 @@ function lastCompoundSelectorWithoutPseudoClasses(selectorNode) {
...
 }
 
+/**
+ * Specificity only has an effect on declarations.
+ * We only want to check rules that contain declarations either directly or in nested at-rules.
+ *
+ * @param {import('postcss').Container} node
+ * @returns {boolean}
+ */
+function nodeContainsDeclarations(node) {
```
Comment (Mouvedia) [inline lib/rules/no-descending-specificity/index.mjs:167]: nit
style

`hasDeclaration`
i.e. only one is enough

File: lib/rules/no-descending-specificity/index.mjs [hunk 2]
```diff
@@ -151,6 +157,21 @@ function lastCompoundSelectorWithoutPseudoClasses(selectorNode) {
...
+ * We only want to check rules that contain declarations either directly or in nested at-rules.
+ *
+ * @param {import('postcss').Container} node
+ * @returns {boolean}
+ */
+function hasDeclaration(node) {
+	if (!node.nodes) return false;
+
+	return node.some((child) => {
+		return isDeclaration(child) || (isAtRule(child) && hasDeclaration(child));
```
Comment (Mouvedia) [inline lib/rules/no-descending-specificity/index.mjs:171]: question
minor

Is `from { margin-top: 50px; }` considered a declaration?
i.e. inside an `@keyframes`

File: lib/rules/no-descending-specificity/__tests__/index.mjs
Comment (Mouvedia) [inline lib/rules/no-descending-specificity/__tests__/index.mjs:1]: minor
non-blocking

I thought of a weird edge case: rule-less layers.
i.e. `a { @layer foo; }`
Could you add skipped test with a `description` to show its not supported?
e.g. `description: not supported because it's useless`

File: lib/rules/no-descending-specificity/index.mjs (hunk 2 above)
Comment (romainmenke) [reply lib/rules/no-descending-specificity/index.mjs:171]: Yes, this is a declaration

File: lib/rules/no-descending-specificity/__tests__/index.mjs
Comment (romainmenke) [reply lib/rules/no-descending-specificity/__tests__/index.mjs:1]: I don't understand.
Why would we add a skipped test for this, why is it not supported?

Comment (Mouvedia) [reply lib/rules/no-descending-specificity/__tests__/index.mjs:1]: So `foo;` is also considered a declaration? Interesting.
I would have thought it would be an identifier list.

ref: https://drafts.csswg.org/css-cascade-5/#typedef-layer-name

Comment (romainmenke) [reply lib/rules/no-descending-specificity/__tests__/index.mjs:1]: No, it isn't a declaration. It's an at rule.

But I don't understand why you say it isn't supported or why you think it is useless :)

Name defining at rules loke `@layer` aren't affected by specificity.

There is no difference between these:

```css
@layer foo;

a {}
```

```css
a { @layer foo; }
```

```css
#a { @layer foo; }
```

Naming defining at rules are only affected by source order and cascade layers themselves.

Comment (Mouvedia) [reply lib/rules/no-descending-specificity/__tests__/index.mjs:1]: > No, it isn't a declaration. It's an at rule.

So `(isAtRule(child) && hasDeclaration(child)` returns `false` and it was intended.

Comment (romainmenke) [reply lib/rules/no-descending-specificity/__tests__/index.mjs:1]: Indeed :)

Currently the CSS specification does not have any at rules that define things (like `@layer foo;` defines a layer named `foo`) and where the definition is affected by specificity of parent rules.

This is unlikely (but not impossible) to change in the future.

It is extremely complex for browsers to implement named things like keyframes, cascade layers, fonts, that are effected by matched elements and specificity.

**Review by ybiquitous — APPROVED** [review]
Thank you, nice fix 👍🏼 

`hasDeclaration()` could be put in `lib/utils/` if it will be used by other rules. 😃
